File: python/ops.py
import trees
import uuid
import json

def safe_bound(x):
    if x > 0:
        return x
    else:
        return 0


### Base Operation.
class Operation(object):
    for_type = '*'

    def __init__(self):
        self._id = str(uuid.uuid4())
        #for the new way of thinking
        self.target_revision = None

    def __repr__(self):
        return "<Op: %s>" % self.pack()

    @classmethod
    def unpack(cls, args):
        klass = op_lookup[args[0]]
        if len(args) > 3 and type(args[3]) == list:
            args[3] = map(lambda x: trees.Node.unpack(x), args[3])

        k = klass(*args[2:])
        k._id = args[1]
        return k

    def clone(self):
        # This is a all around bad method, but frankly, I want to ensure that
        # if you get an object it can't change out from under you, so this is
        # one way to get it done until I come up with a cleaner way.
        return Operation.unpack(self.pack())

    ###### Now it gets fun, this should eventually be split out into classes or
    ###### some sort of lookup table so that the user could feasibly define
    ###### their own, although, that also seems like a bad idea.
    def handle_mutate(self, root, tpath, oplist):
        new_list = []
        for row in oplist:
            new_row = row

            if tpath and tpath != row[1]:
                if all([i == j for i, j in zip(tpath, row[1])]):
                    new_row = self.handle_mutate_with_path(row, tpath)
                new_list.append(new_row)
            else:
                if self.for_type == 'dict':
                    if self.source_path == row[1]:
                        new_row = self.handle_mutate_without_path(row, tpath)
                        root = root.get_path(new_row[1])
                elif self.for_type in ('list', 'string'):
                    if self.index > row[2].index:
                        ## this means that modifications don't affect
                        ## this op.
                        pass
                    else:
                        new_row = self.handle_mutate_without_path(row, tpath)
                else:
                    raise Exception('unsupported type')

                newroot, newrev = new_row[2].apply(root)
                new_list.append([new_row[0], new_row[1], new_row[2], newrev])
                root = newroot

        return new_list

    def handle_mutate_with_path(self, row, tpath):
        if self.for_type == 'string':
            return self.handle_mutate_with_path_string(row, tpath)
        elif self.for_type == 'list':
            return self.handle_mutate_with_path_list(row, tpath)
        elif self.for_type == 'dict':
            return self.handle_mutate_with_path_dict(row, tpath)
        else:
            return row

    def handle_mutate_with_path_string(self, row, tpath):
        p = row[2].clone()
        if type(self) == StringInsertOperation:
            p.index = safe_bound(p.index + len(self.text))
        elif type(self) == StringDeleteOperation:
            p.index = safe_bound(p.index - self.length)
        return [row[0], row[1], p, row[3]]

    def handle_mutate_with_path_list(self, row, tpath):
        path = row[1]
        if type(self) == ListInsertOperation:
            path[len(tpath)] = safe_bound(path[len(tpath)] + len(self.value))
        elif type(self) == ListDeleteOperation:
            path[len(tpath)] = safe_bound(path[len(tpath)] - self.length)

        return [row[0], path, row[2], row[3]]

    def handle_mutate_without_path(self, row, tpath):
        if self.for_type == 'string':
            return self.handle_mutate_without_path_string(row, tpath)
        elif self.for_type == 'list':
            return self.handle_mutate_without_path_list(row, tpath)
        elif self.for_type == 'dict':
            return self.handle_mutate_without_path_dict(row, tpath)
        else:
            return row

    def handle_mutate_without_path_dict(self, row, tpath):
        return [row[0], self.target_path, row[2], row[3]]
        
    def handle_mutate_without_path_string(self, row, tpath):
        p = row[2].clone()

        if type(self) == StringInsertOperation:
            p.index = safe_bound(p.index + len(self.text))
        elif type(self) == StringDeleteOperation:
            p.index = safe_bound(p.index - self.length)

        return [row[0], row[1], p, row[3]]

    def handle_mutate_without_path_list(self, row, tpath):
        p = row[2].clone()
        if type(self) == ListInsertOperation:
            p.index = safe_bound(p.index + len(self.value))
        elif type(self) == ListDeleteOperation:
            p.index = safe_bound(p.index - self.length)
        return [row[0], row[1], p, row[3]]
    ### End of GOT type conflict resolution.

    ### Begin Pavelian History Maintenence.
    def cross_merge(self, previous_transform):
        assert self.for_type == previous_transform.for_type

        if self.for_type in trees.APPLY_TYPES:
            return self
        elif self.for_type in (trees.TYPES['string'], trees.TYPES['list']):
            pass
    ### End Pavelian History Maintenence.


### Basic Set Operation
class SetOperation(Operation):
    for_type = '*'

    def __init__(self, value):
        Operation.__init__(self)
        self.value = value

    def pack(self):
        return ['SetOperation', self._id, self.value]

    def apply(self, node):
        new = trees.Node.from_obj(self.value)
        reverse = SetOperation(node.obj_repr())
        return new, reverse


### Number Operations.
class NumberIncrementOperation(Operation):
    for_type = 'number'

    def __init__(self, amount):
        Operation.__init__(self)
        self.amount = amount

    def pack(self):
        return ['NumberIncrementOperation', self._id, self.amount]

    def apply(self, node):
        new = node.clone()
        new.set_value(node.obj_repr() + self.amount)
        reverse = NumberIncrementOperation(self.amount * (-1))
        return new, reverse


### String Operations.
class StringInsertOperation(Operation):
    for_type = 'string'

    def __init__(self, index, text):
        Operation.__init__(self)
        self.index = index
        self.text = text

    def pack(self):
        return ['StringInsertOperation', self._id, self.index, self.text]

    def apply(self, node):
        node_value = str(node.obj_repr())
        nv = node_value[:self.index] + self.text + node_value[self.index:]
        new = node.clone()
        new.set_value(nv)

        reverse = StringDeleteOperation(self.index, len(self.text))
        return new, reverse


class StringDeleteOperation(Operation):
    for_type = 'string'

    def __init__(self, index, length):
        Operation.__init__(self)
        self.index = index
        self.length = length

    def pack(self):
        return ['StringDeleteOperation', self._id, self.index, self.length]

    def apply(self, node):
        node_value = str(node.obj_repr())
        new = node.clone()
        new.set_value(node_value[:self.index] +
                      node_value[self.index + self.length:])

        reverse = StringInsertOperation(
            self.index, node_value[self.index:self.index + self.length])
        return new, reverse


class StringSetOperation(Operation):
    for_type = 'string'

    def __init__(self, value):
        Operation.__init__(self)
        self.value = value

    def pack(self):
        return ['StringSetOperation', self._id, self.value]

    def apply(self, node):
        new = node.clone()
        new.set_value(self.value)
        reverse = StringSetOperation(node.obj_repr())
        return new, reverse


### Boolean operations.
class BooleanSetOperation(Operation):
    for_type = 'boolean'

    def __init__(self, value):
        Operation.__init__(self)
        self.value = value

    def pack(self):
        return ["BooleanSetOperation", self._id, self.value]

    def apply(self, node):
        new = node.clone()
        new.set_value(self.value)

        reverse = BooleanSetOperation(not self.value)
        return new, reverse


### List operations.
class ListInsertOperation(Operation):
    for_type = 'list'

    def __init__(self, index, value):
        Operation.__init__(self)
        self.index = index
        self.value = value

    def pack(self):
        return ["ListInsertOperation", self._id, self.index, [i.pack() for i in self.value]]

    def apply(self, node):
        new = node.clone()
        new.children = new.children[:self.index] + self.value + new.children[self.index:]
        reverse = ListDeleteOperation(self.index, len(self.value))
        return new, reverse


class ListDeleteOperation(Operation):
    for_type = 'list'

    def __init__(self, index, length):
        Operation.__init__(self)
        self.index = index
        self.length = length

    def pack(self):
        return ["ListDeleteOperation", self._id, self.index, self.length]

    def apply(self, node):
        new = node.clone()
        reverse = ListInsertOperation(
            self.index, new.children[self.index:self.index + self.length])
        new.children = new.children[:self.index] + new.children[self.index + self.length:]
        return new, reverse


class ListSetIndexOperation(Operation):
    for_type = 'list'

    def __init__(self, index, value):
        Operation.__init__(self)
        self.index = index
        self.value = value

    def pack(self):
        return ["ListSetIndexOperation", self._id, self.index, self.value]

    def apply(self, node):
        old = node.children[self.index]
        new_node = node.clone()
        new_node.children = []

        for index, item in enumerate(node.children):
            if index == self.index:
                new_node.children.append(self.value)
            else:
                new_node.children.append(item)

        reverse = ListSetIndexOperation(self.index, old)
        return new_node, reverse


class ListApplyIndexOperation(Operation):
    for_type = 'list'

    def __init__(self, index, operation):
        Operation.__init__(self)
        self.index = index
        self.operation = operation

    def pack(self):
        return ["ListApplyIndexOperation", self._id, self.index, self.operation.pack()]

    def apply(self, node):
        new, reverse = self.operation.apply(node.children[self.index])
        reverse_op = ListApplyIndexOperation(self.index, reverse)

        new_node = node.clone()
        new_node.children = []

        for index, item in enumerate(node.children):
            if index == self.index:
                new_node.children.append(new)
            else:
                new_node.children.append(item)

        return new_node, reverse_op


### Dictionary Operations.
### This is tricky.
class DictKeyApplyOperation(Operation):
    for_type = 'dict'

    def __init__(self, key, operation):
        Operation.__init__(self)
        self.key = key
        self.operation = operation

    def pack(self):
        return ["DictKeyApplyOperation", self._id, self.key, self.operation.pack()]

    def apply(self, node):
        child = node.get_path(self.key)
        new, reverse = self.operation.apply(child)

        new_node = node.clone()
        new_node.set_path(self.key, new)

        reverse_op = DictKeyApplyOperation(self.key, reverse)
        return new_node, reverse_op

## SUP
class MovePathOperation(Operation):
    for_type = 'dict'

    def __init__(self, source_path, target_path):
        Operation.__init__(self)
        self.source_path = source_path
        self.target_path = target_path

    def pack(self):
        return ['MovePathOperation', self._id, self.source_path, self.target_path]

    def apply(self, node):
        new = node.clone()
        value = new.get_path(self.source_path)
        new.remove_path(self.source_path)
        new.set_path(self.target_path, value)

        reverse = MovePathOperation(self.target_path, self.source_path)
        return new, reverse


#sets
class SetAddOperation(Operation):
    for_type = 'set'

    def __init__(self, value):
        Operation.__init__(self)
        self.value = value

    def pack(self):
        return ["SetAddOperation", self._id, self.value]

    def apply(self, node):
        pass


class SetRemoveOperation(Operation):
    for_type = 'set'

    def __init__(self, value):
        Operation.__init__(self)
        self.value = value

    def pack(self):
        return ["SetRemoveOperation", self._id, self.value]

    def apply(self, node):
        pass


# zsets
class ScoredSetAddOperation(Operation):
    pass


class ScoredSetApplyOperation(Operation):
    pass


class ScoredSetDropOperation(Operation):
    pass


op_lookup = {
    'NumberIncrementOperation': NumberIncrementOperation,
    'StringInsertOperation': StringInsertOperation,
    'StringDeleteOperation': StringDeleteOperation,
    'StringSetOperation': StringSetOperation,
    'BooleanSetOperation': BooleanSetOperation,
    'ListInsertOperation': ListInsertOperation,
    'ListDeleteOperation': ListDeleteOperation,
    'ListSetIndexOperation': ListSetIndexOperation,
    'ListApplyIndexOperation': ListApplyIndexOperation,
    'DictKeyApplyOperation': DictKeyApplyOperation
}

File: js/ops.js
var util = require('util');

var ops = (function() {
        var Operation = function() {};
        Operation.prototype.clone = function() {
            return unpack(this.pack());
        };
        Operation.prototype.toString = function() {
            return "<fuck you>";
        };

        Operation.handle_mutate = function(root, tpath, oplist) {
            var new_list = [];
            for(var index in oplist) {
                var row = oplist[index];
                var new_row = row;

                if (tpath && (tpath != row[1])) {
                    if (true) {}
                }
            }
        };

        // Number Operations.
        var NumberIncrementOperation = function(amount) {
            this.for_type = 'number';
            this.amount = amount;
        };
        NumberIncrementOperation.prototype = new Operation();
        
        NumberIncrementOperation.prototype.apply = function(node) {
            var new_node = node.proto({'value':node.obj_repr() + this.amount});
            var reverse = new NumberIncrementOperation(this.amount * (-1));
            return [new_node, reverse];
        };
        
        NumberIncrementOperation.prototype.pack = function() {
            return ["NumberIncrementOperation", this.amount];
        };

        // end Number Operations.

        // String Operations
        var StringInsertOperation = function(index, text) {
            this.for_type = 'string';
            this.index = index;
            this.text = text;
        };
        StringInsertOperation.prototype = new Operation();

        StringInsertOperation.prototype.apply = function(node) {
            var node_value = node.obj_repr();
            var new_node = node.proto({'value':node_value.slice(0, this.index) + this.text + node_value.slice(this.index, node_value.length)});
            var reverse = new StringDeleteOperation(this.index, this.text.length);
            return [new_node, reverse];
        };

        StringInsertOperation.prototype.pack = function() {
            return ["StringInsertOperation", this.index, this.text];
        };

        var StringDeleteOperation = function(index, length) {
            this.for_type = 'string';
            this.index = index;
            this.length = length;
        };
        StringDeleteOperation.prototype = new Operation();

        StringDeleteOperation.prototype.apply = function(node) {
            var node_value = node.obj_repr();
            var new_node = node.proto({'value':node_value.slice(0, this.index) + node_value.slice(this.index + this.length, node_value.length)});
            var reverse = new StringInsertOperation(this.index, node_value.slice(this.index, this.length));
            return [new_node, reverse];
        };

        StringDeleteOperation.prototype.pack = function() {
            return ["StringDeleteOperation", this.index, this.length];
        };

        var StringSetOperation = function(value) {
            this.for_type = 'string';
            this.value = value;
        };
        StringSetOperation.prototype = new Operation();

        StringSetOperation.prototype.apply = function(node) {
            var new_node = node.proto({'value':this.value});
            var reverse = new StringSetOperation(node.obj_repr());
            return [new_node, reverse];
        };

        StringSetOperation.prototype.pack = function() {
            return ["StringSetOperation", this.value];
        };
        // end String Operations

        // Boolean Operations
        var BooleanSetOperation = function(value) {
            this.for_type = 'boolean';
            this.value = value;
        };
        BooleanSetOperation.prototype = new Operation();

        BooleanSetOperation.prototype.apply = function(node) {
            var new_node = node.proto({'value':this.value});
            var reverse = new BooleanSetOperation(!this.value);
            return [new_node, reverse];
        };

        BooleanSetOperation.prototype.pack = function() {
            return ["BooleanSetOperation", this.value];
        };
        // end Boolean Operations

        // List Operations
        var ListInsertOperation = function(index, value) {
            this.for_type = 'list';
            this.index = index;
            this.value = value;
        };
        ListInsertOperation.prototype = new Operation();

        ListInsertOperation.prototype.apply = function(node) {
            var new_node = node.proto({
                    'children':
                    node.children.slice(0, this.index)
                    .concat(this.value)
                    .concat(node.children.slice(this.index, node.children.length))
                });
            var reverse = new ListDeleteOperation(this.index, this.value.length);
            return [new_node, reverse];
        };

        ListInsertOperation.prototype.pack = function() {
            return ["ListInsertOperation", this.index, this.value];
        };

        var ListDeleteOperation = function(index, length) {
            this.for_type = 'list';
            this.index = index;
            this.length = length;
        };
        ListDeleteOperation.prototype = new Operation();

        ListDeleteOperation.prototype.apply = function(node) {
            var new_node = node.proto({
                    'children':
                    node.children.slice(0, this.index)
                    .concat(node.children.slice(this.index+1, node.children.length))
                });
            var reverse = new ListInsertOperation(this.index, node.children.slice(this.index, this.length));
            return [new_node, reverse];
        };

        ListDeleteOperation.prototype.pack = function() {
            return ["ListDeleteOperation", this.index, this.length];
        };

        var ListSetIndexOperation = function(index, value) {
            this.for_type = 'list';
            this.index = index;
            this.value = value;
        };
        ListSetIndexOperation.prototype = new Operation();

        ListSetIndexOperation.prototype.apply = function(node) {
            var old = node.children[this.index];
            var new_node = node.proto({'value':node.value, 'children':[]});
            for(var index in node.children) {
                if (index == this.index) {
                    new_node.children.push(this.value);
                } else {
                    new_node.children.push(node.children[index]);
                }
            }
            var reverse = new ListSetIndexOperation(this.index, old);
            return [new_node, reverse];
        };

        ListSetIndexOperation.prototype.pack = function() {
            return ["ListSetIndexOperation", this.index, this.value];
        };

        var ListApplyIndexOperation = function(index, operation) {
            this.for_type = 'list';
            this.index = index;
            this.operation = operation;
        };
        ListApplyIndexOperation.prototype = new Operation();

        ListApplyIndexOperation.prototype.apply = function(node) {
            var nr = this.operation.apply(node.children[this.index]);
            var reverse_op = new ListApplyIndexOperation(this.index, nr[1]);
            var new_node = node.proto({'value':node.value, 'children':[]});
            for(var index in node.children) {
                if (index == this.index) {
                    new_node.children.push(nr[0]);
                } else {
                    new_node.children.push(node.children[index]);
                }
            }
            return [new_node, reverse_op];
        };

        ListApplyIndexOperation.prototype.pack = function() {
            return ["ListApplyIndexOperation", this.index, this.operation.pack()];
        };
        // end List Operations
        
        // Dictionary Key Apply
        var DictKeyApplyOperation = function(key, operation) {
            this.for_type = 'dict';
            this.key = key;
            this.operation = operation;
        };
        DictKeyApplyOperation.prototype = new Operation();

        DictKeyApplyOperation.prototype.apply = function(node) {
            var child = node.get_path(this.key);
            var nr = this.operation.apply(child);
            var new_node = node.proto({'value':node.value, 'children':[]});
            for(var index in node.children) {
                new_node.children.push(node.children[index]);
            }
            new_node.set_path(this.key, nr[0]);
            reverse_op = new DictKeyApplyOperation(this.key, nr[1]);
            return [new_node, reverse_op];
        };

        DictKeyApplyOperation.prototype.pack = function() {
            return ["DictKeyApplyOperation", this.key, this.operation.pack()];
        };
        // end dict

        //class MoveBookmarkOperation(Operation): 
        //class CopyBookmarkOperation(Operation):
        //class PruneBookmarkOperation(Operation):

        var exports = {
            'NumberIncrementOperation':NumberIncrementOperation,
            'StringInsertOperation':StringInsertOperation,
            'StringDeleteOperation':StringDeleteOperation,
            'StringSetOperation':StringSetOperation,
            'BooleanSetOperation':BooleanSetOperation,
            'ListInsertOperation':ListInsertOperation,
            'ListDeleteOperation':ListDeleteOperation,
            'ListSetIndexOperation':ListSetIndexOperation,
            'ListApplyIndexOperation':ListApplyIndexOperation,
            'DictKeyApplyOperation':DictKeyApplyOperation
        };

        var unpack = function(d) {
            // there is some trickery here, but only to make things easier.
            var klass = d[0];
            var obj = new exports[klass].prototype.constructor();
            exports[klass].prototype.constructor.apply(obj, d.slice(1));
            return obj;
        };

        exports['unpack'] = unpack;

        // Lets see if we are running in the Node.js context or in a browser.
        if (typeof module !== 'undefined') {
            module.exports = exports;
        }
        return exports;
    })();
